mixer/cmd/reset.go

```diff
@@ -0,0 +1,268 @@
+// Copyright © 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/clearlinux/mixer-tools/swupd"
+
+	"github.com/clearlinux/mixer-tools/builder"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+var resetCmd = &cobra.Command{
+	Use:   "reset",
+	Short: "Revert mix to a previous version",
+	Long: `Revert the mix to a previous version.
+Reverts a mix state to the end of the last build 
+or to the end of a given version build if one is provided. 
+By default, the value of PREVIOUS_MIX_VERSION 
+in mixer.state will be used to define the last build. 
+This command can be used to roll back the mixer state 
+in case of a build failure or in case the user 
+wants to roll back to a previous version.`,
+	RunE: runReset,
+}
+
+type resetFlags struct {
+	toVersion int32
+	clean     bool
+}
+
+var resetCmdFlags resetFlags
+
+func init() {
+	RootCmd.AddCommand(resetCmd)
+
+	resetCmd.Flags().Int32Var(&resetCmdFlags.toVersion, "to", -1, "Reset to a specific mix version, default = PREVIOUS_MIX_VERSION")
+	resetCmd.Flags().BoolVar(&resetCmdFlags.clean, "clean", false, "Deletes all files with versions bigger than the one provided")
+}
+
+func runReset(cmd *cobra.Command, args []string) error {
+	b, err := builder.NewFromConfig(configFile)
+	if err != nil {
+		return err
+	}
+
+	// if toVersion provided by the user, replace the mixer version
+	if resetCmdFlags.toVersion >= 0 {
+		b.MixVer = strconv.Itoa(int(resetCmdFlags.toVersion))
+		b.MixVerUint32 = uint32(resetCmdFlags.toVersion)
+	} else {
+		// assuming mixer.state file has the correct info
+		b.MixVer = b.State.Mix.PreviousMixVer
+		b.MixVerUint32, err = parseUint32(b.State.Mix.PreviousMixVer)
+		if err != nil {
+			return err
+		}
+	}
+
+	if b.MixVer == "0"{
```

Comment:
mixversions are required to be > 0. Allowing a mix to be reset to 0 will lead to failures later.

```
mixer init
sudo mixer reset
sudo mixer build all
ERROR: Couldn't build update: failed to create update metadata: manifest has version zero, version must be positive
```

Reply:
yes, I will make it invalid to give reset 0 then

Reply:
what will be good approach to reset the mix to a state which is after mixer init?

Reply:
If the previous version is 0, I would use the current mixversion. If that is invalid (< 1), I would print a warning and default to 10 which is the `mixer init` mixversion default.

Reply:
@jwakre  can you test the recent commit. I made the changes.